### ISSUE TITLE
[Feature Request] Support download latest driver version

### ISSUE THREAD

**Issue by trinhpham** [vvo/selenium-standalone#468]
## Expected Behavior
Basically, automation scripts should know the exact version of browser & browser's driver it is going to interact.
However, browsers are updated automatically, there's a need to get our automation scripts be updated.
I think it would be the best if we can set `LATEST` to the `driver.version` at the installation phase, then the service would install latest drivers version.

## Current Behavior
We have to specify the correct version, or the hard-coded version in `lib\default-config.js` will be installed.

## Action items
IMO, I'd like to propose that we should rollout the support one by one:
- [x] **Selenium server**
I haven't researched how to get the latest version yet. It seems that the Selenium team has not released new stable version for long time. So should we assume `3.141.59` is the latest version?

- [x] **ChromeDriver** 
We can check content of [this file](https://chromedriver.storage.googleapis.com/LATEST_RELEASE) to know the latest version.

- [ ] **GeckoDriver**
We can read the redirect URL of this link to know the latest release tag

- [x] **Legacy Microsoft Edge**
I think Microsoft won't release more version of this (Am I correct?). So should we assume `17134` is the latest version?

- [x] **Chromium-based Microsoft Edge**
The same to ChromeDriver, we can read the content of [this file](https://msedgewebdriverstorage.blob.core.windows.net/edgewebdriver/LATEST_STABLE) to find the latest version.

How are your thoughts? Should we make it happen? Where should we start? :)

**Comment by vvo** [comment]
> * I haven't researched how to get the latest version yet. It seems that the Selenium team has not released new stable version for long time. So should we assume `3.141.59` is the latest version?

I guess you could also parse https://selenium-release.storage.googleapis.com/, find the most recent date and use that.

> * I think Microsoft won't release more version of this (Am I correct?)

No idea! But you could always ask them/their maintainers if you hunt them down or find the right repository to ask the question.

> How are your thoughts? Should we make it happen?

Yes, I think it's a good idea, it could be the new default behavior if it makes sense (given we bump a major version). Thus we could remove the "default config" file and only rely on the versions you will detect.

> Where should we start? :)
Have a look at the code, use the entry point and read how the code works. Try to map the program workflow in your mind and start with that.

You can easily develop this locally:
- fork the repository
- npm install
- npm link # this should install a globally "selenium-standalone" bin that will be your local version
- modify the files, run selenium-standalone manually and see how it goes from there


I guess one hacky/quick way to implement this could be to update this code: https://github.com/vvo/selenium-standalone/blob/f73b1a96b899f6c4bed95777611fe5b099b54c80/bin/selenium-standalone#L55, so that "defaultConfig" is computed instead of just read from a file. Something like:

```js
const defaultConfig = await computeDefaultConfig(); // computeDefaultConfig would be a function that basically does a Promise.map(findAllLatestVersions) in parallel.
```

The issue might be that for now this project supports Node.js versions that might not support await (I actually don't know exactly if that's the case), so you'd have to use callbacks instead/promises and wraps the code differently.

Good luck!

**Comment by trinhpham** [comment]
> The issue might be that for now this project supports Node.js versions that might not support await (I actually don't know exactly if that's the case), so you'd have to use callbacks instead/promises and wraps the code differently.

Due to [NodeJS release], even NodeJS v8 - the version that support promises/await - was retired. Should we drop support all NodeJS versions that are older than 8 @vvo ?

**Comment by vvo** [comment]
> Should we drop support all NodeJS versions that are older than 8 @vvo ?

Yes, that should be fine indeed, see https://nodejs.org/metrics/#operating-systems

**Comment by frathe** [comment]
This should be quite high on the agenda.
Finding the correct way to configure this in the codeceptjs config was painful, since there is no documentation of this.
I just leave this here should anybody else fall into this pit:
```
exports.config = {
    helpers: {
        WebDriver: {
            seleniumArgs: {
                drivers: {
                    chrome: {
                        version: '85.0.4183.83',
                        arch: process.arch,
                        baseURL: 'https://chromedriver.storage.googleapis.com'
                    },
                }
            }
        },
    }
};
```

**Comment by trinhpham** [comment]
@vvo: I found that [webdriver-manager](https://github.com/angular/webdriver-manager) is doing the job already. It's actively maintain by some people there. It's sad to say that this package has completed its mission really well and can be retired. I'll leave this feature request open so if you (and others) would like to keep going on.

@frathe CodeceptJS uses WDIO behind the scene, so you can find the [document at wdio](https://webdriver.io/docs/selenium-standalone-service.html). However, it's not recommended by CodeceptJS if you are planning to run your tests in parallel. Below is a quote [from CodeceptJS](https://codecept.io/webdriver/#what-is-selenium-webdriver)
> ⚠ It is not recommended to use wdio plugin & selenium-standalone when running tests in parallel. Consider switching to Selenoid if you need parallel run or using cloud services.

**Comment by vvo** [comment]
@trinhpham thanks! I would be more than happy to archive this repository and point to webdriver-manager. From your experience, are every features from selenium-standalone inside webdriver-manager?

Also pinging @christian-bromann, hi there, do you know if we could replace selenium-standalone completely with https://github.com/angular/webdriver-manager? thanks!

**Comment by trinhpham** [comment]
@vvo not all, but it matches some of my need:

- [x] Able to get latest & specified driver version of Chrome & Firefox
- [x] Able to get latest & specified jar file version of selenium 
- [x] Able to start the selenium server with downloaded drivers
- [ ] WDIO plugin
- [ ] MS Edge Chromium

**Comment by christian-bromann** [comment]
> It's actively maintain by some people there.

It hasn't seen any contributions for over a year.

> do you know if we could replace selenium-standalone completely with [angular/webdriver-manager](https://github.com/angular/webdriver-manager)? thanks!

Theoretically you could replace it without the user seeing much difference.

**Comment by mgrybyk** [comment]
done for chrome and edge chromium. I've raised another ticket to deal with geckodriver in the future.

Feel free to raise for issue/PR other components/drivers if needed.

Thanks
